rich/live.py

```diff
@@ -0,0 +1,364 @@
+import sys
+from threading import Event, RLock, Thread
+from typing import IO, Any, List, Optional
+
+from typing_extensions import Literal
+
+from .__init__ import get_console
+from ._loop import loop_last
+from .console import (
+    Console,
+    ConsoleOptions,
+    ConsoleRenderable,
+    RenderableType,
+    RenderHook,
+    RenderResult,
+)
+from .control import Control
+from .jupyter import JupyterMixin
+from .live_render import LiveRender
+from .progress import _FileProxy
+from .segment import Segment
+from .style import Style
+
+VerticalOverflowMethod = Literal["crop", "ellipsis", "visible"]
+
+
+class _RefreshThread(Thread):
+    """A thread that calls refresh() at regular intervals."""
+
+    def __init__(self, live: "Live", refresh_per_second: float) -> None:
+        self.live = live
+        self.refresh_per_second = refresh_per_second
+        self.done = Event()
+        super().__init__()
+
+    def stop(self) -> None:
+        self.done.set()
+
+    def run(self) -> None:
+        while not self.done.wait(1 / self.refresh_per_second):
+            with self.live._lock:
+                self.live.refresh()
+
+
+class _LiveRender(LiveRender):
+    def __init__(self, live: "Live", renderable: RenderableType) -> None:
+        self._live = live
+        self.renderable = renderable
+        self._shape: Optional[Tuple[int, int]] = None
+
+    def __rich_console__(
+        self, console: Console, options: ConsoleOptions
+    ) -> RenderResult:
+        with self._live._lock:
+            lines = console.render_lines(self.renderable, options, pad=False)
+
+            shape = Segment.get_shape(lines)
+            _, height = shape
+            if height > console.size.height:
+                if self._live.vertical_overflow == "crop":
+                    lines = lines[: console.size.height]
+                    shape = Segment.get_shape(lines)
+                elif self._live.vertical_overflow == "ellipsis":
+                    lines = lines[: (console.size.height - 1)] + [
```

Comment:
Would this put the ellipsis on the left hand side of the terminal? I suspect it would look better centered. Also, what would happen if the terminal is less than three characters in width (crazy I know, just covering all the bases).

Reply:
Alright I added in the console render suggestion to cover this.